packages/api-contract/src/MetaRegistry.ts

```diff
@@ -0,0 +1,333 @@
+
+import * as t from '@polkadot/types/types';
+
+import { assert } from '@polkadot/util';
+import { encodeEnum, encodeStruct, encodeTuple } from '@polkadot/types';
+
+const SPECIAL_TYPES = ['AccountId', 'Balance'];
+
+export default class MetaRegistry {
+  private _strings: string[] = [];
+
+  private _types: t.MetaType[] = [];
+
+  public typeDefs: t.TypeDef[] = [];
+
+  public constructor ({ registry: { strings, types } }: t.MetaRegistryJson) {
+    this._strings = strings;
+    this._types = types;
+
+    this._types.forEach((_, index: number): void => this.setTypeDefAtIndex(index + 1));
+  }
+
+  public stringAt (index: t.StringIndex): string {
+    assert(this._strings && index <= this._strings.length, `MetaRegistry: Invalid string index ${index} found in metadata`);
+
+    return this._strings[index - 1];
+  }
+
+  public stringsAt (indices: t.StringIndex[]): string[] {
+    return indices.map(
+      (index: number): string => {
+        return this.stringAt(index);
+      }
+    );
+  }
+
+  public typeAt (index: t.TypeIndex): t.MetaType {
+    assert(this._types && index <= this._types.length, `MetaRegistry: Invalid type index ${index} found in metadata`);
+
+    return this._types[index - 1];
+  }
+
+  public typesAt (indices: t.TypeIndex[]): t.MetaType[] {
+    return indices.map(
+      (index: number): t.MetaType => {
+        return this.typeAt(index);
+      }
+    );
+  }
+
+  public hasTypeDefAt (index: t.TypeIndex): boolean {
+    return !!this.typeDefs[index - 1];
+  }
+
+  public typeDefAt (index: t.TypeIndex, extra: Pick<t.TypeDef, never> = {}): t.TypeDef {
+    assert(this.hasTypeDefAt(index), `Invalid TypeDef index requested: ${index}`);
+    return {
+      ...this.typeDefs[index - 1],
+      ...extra
+    };
+  }
+
+  private typeDefIdFields ({ id }: t.MetaType): Pick<t.TypeDef, never> {
+    if (!id) {
+      return {};
+    }
+
+    const {
+      'custom.name': nameIndex,
+      'custom.namespace': namespaceIndices,
+      'custom.params': paramsIndices
+    } = id as t.MetaTypeIdCustom;
+
+    if (nameIndex) {
+      return {
+        name: this.stringAt(nameIndex),
+        ...(
+          namespaceIndices && namespaceIndices.length
+            ? { namespace: this.stringsAt(namespaceIndices).join('::') }
+            : {}
+        ),
+        ...(
+          paramsIndices && paramsIndices.length
+            ? { params: this.typesAt(paramsIndices).map((type): t.TypeDef => this.typeDefFromMetaType(type)) }
+            : {}
+        ),
+        ...(SPECIAL_TYPES.includes(this.stringAt(nameIndex))
+          ? { type: this.stringAt(nameIndex) }
+          : {}
+        )
+      };
+    }
+
+    return {};
+  }
+
+  private typeDefDefFields (metaType: t.MetaType, typeIndex?: t.TypeIndex): Pick<t.TypeDef, never> {
+    assert(!(metaType.def as t.MetaTypeDefUnion)['union.fields'], 'Invalid union type definition found');
+
+    let typeDef: Pick<t.TypeDef, never> = {};
```

Comment:
Just `let typeDef: Pick<t.TypeDef, never>`;

(Then remove the `assert` below and make a final leg with `else` after the last `else if` where we throw.